HPCC-18294 dfuserver segfault

AttilaVamos opened the pull request:
Check current file part pointer validity before call Release()

Fix some code formatting issues in FileSizeThread::wait() and
FileSizeThread::run() functions.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually, however it is hard to reproduce environment to create similar circumstances reported in JIRA.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

AttilaVamos commented:
@jakesmith please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: d1d6bc2bde6e37b45105692dec17f32b3c5753da
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 93.832 seconds |


Install hpccsystems-platform-community_6.4.3-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented on dali/ft/filecopy.cpp:
```diff
@@ -465,29 +465,34 @@ FileSizeThread::FileSizeThread(FilePartInfoArray & _queue, CriticalSection & _cs
             StringBuffer url;
             WARNLOG("Waiting for file: %s",rfn->getRemotePath(url).str());
-            cur->Release();
+            if (cur.get())
```
This looks like the change you made, but if cur is null it will have crashed earlier on the cur->link() I would have thought (but then again, if cur is null then the whole loop should have been skipped by the continue).

If some other thread is modifying cur concurrently, then this code has bigger problems than you can solve with a simple check for non-null.

jakesmith replied:
Agree, it shouldn't be here if cur was null, so this additional check should not be needed/helpful.
The fact it is null (backtrace suggests it is) at this point, suggests there's a race condition or overlapping thread, that's clearing cur, after the if !cur.get() test.

AttilaVamos replied:
I saw in my attempts to reproduce this problem the value of cur can changed by FileSizeThread::run() during the WARNLOG(...) finished.

jakesmith replied:
This code is essentially not thread safe afaics.
At any point during FileSizeThread::wait, cur could be cleared, so the crash seen was one of a number of possible crashes.
The fix should ensure/protect the object FileSizeThread::wait is dealing with cannot be altered, e.g. by using a CS lock whilst it a link copy of cur.

jakesmith commented:
@AttilaVamos - back to you.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: f6ad48a8363a20878f22a9fa4d61c0cf884ca64a
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.228 seconds |


Install hpccsystems-platform-community_6.4.3-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

AttilaVamos commented:
@jakesmith please take a look.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -465,29 +465,32 @@ FileSizeThread::FileSizeThread(FilePartInfoArray & _queue, CriticalSection & _cs
+            }
         }
-        if (rfn) {
+        cs.leave();
```
Should use CriticalBlock b(cs), rather than separate enter(), leave() calls in general, because cleaner and ensures leave() happens.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -465,29 +465,32 @@ FileSizeThread::FileSizeThread(FilePartInfoArray & _queue, CriticalSection & _cs
         }
-        if (rfn) {
+        cs.leave();
+        if (rfn)
```
This is out of the crit now, cur could have been destroyed and rfn is a pointer to cur->mirrorFilename...
So this still has similar problems that caused the original crash, i.e. not thread safe.

jakesmith commented:
The handling of 'cur' still needs some thought to make it thread safe.
FileSizeThread::run() alters cur and it's contents outside of the critical section, and wait() uses it's content outside, e.g.:
+ cur.clear() - near top of run() , and 
+ 'if (copy==1)' - block near the bottom, which alters it's contents.

@AttilaVamos

AttilaVamos commented:
@jakesmith new attempt, please take a look.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: a7e9a6396407e00ddc82ea8eb89acd7cd74e8c1b
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.132 seconds |


Install hpccsystems-platform-community_6.4.4-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -536,7 +546,8 @@ int FileSizeThread::run()
                 break;
             }
-            if (copy==1) { // need to set primary
+            if (copy==1)
```
The code below alters then potentially destroys cur, the other thread in wait() could be accessing it.
This needs protecting too.

jakesmith reviewed (changes requested):
Added some comments.
The issues previously mentioned here : https://github.com/hpcc-systems/HPCC-Platform/pull/10450#issuecomment-335840653

still need addressing.

jakesmith commented:
@AttilaVamos - back to you.

AttilaVamos commented:
@jakesmith next iteration, please check.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: c76891cbc3cf3ef80b66473fb49ac502da38294e
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.247 seconds |


Install hpccsystems-platform-community_6.4.4-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -465,29 +465,34 @@ FileSizeThread::FileSizeThread(FilePartInfoArray & _queue, CriticalSection & _cs
+                if (copy)
+                {
+                    if (!cur->mirrorFilename.isNull())
+                        rfn.set(cur->mirrorFilename.getRemotePath(url));
```
you don't need both 'rfn' and 'utl' StringBuffer's.
You can just pass 'rfn' to getRemoteBuffer only, i.e.:
```
   cur->mirrorFilename.getRemotePath(rn));
```

and remove line 473

AttilaVamos replied:
True, changed.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -465,29 +465,34 @@ FileSizeThread::FileSizeThread(FilePartInfoArray & _queue, CriticalSection & _cs
+                }
+                else
+                {
+                    rfn.set(cur->filename.getRemotePath(url));
```
Same as above - pass the 'rfn' StringBuffer to getRemotePath.

AttilaVamos replied:
Changed.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -465,29 +465,34 @@ FileSizeThread::FileSizeThread(FilePartInfoArray & _queue, CriticalSection & _cs
+        }
+        if (!rfn.isEmpty())
+        {
+            StringBuffer url;
```
unused, should be removed.

AttilaVamos replied:
Removed.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -497,17 +502,20 @@ int FileSizeThread::run()
-                cur.setown(&queue.popGet());
-            cs.leave();
+            {
+                cur.clear();
```
The cur.clear() is still a problem, as it's outside the crit.

AttilaVamos replied:
Moved into.

jakesmith commented:
@AttilaVamos - back to you.

AttilaVamos commented:
@jakesmith Updated, please check.

jakesmith commented:
ok don't see any further issues.
Can you squash the commits now? @AttilaVamos

AttilaVamos committed:
HPCC-18294 dfuserver segfault

Put data member access in critical sections and remove Link() and Release()

Fix some code formatting issues in FileSizeThread::wait() and
FileSizeThread::run() functions.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

AttilaVamos commented:
@jakesmith Squashed.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 506130e761629e3defc983245dbc6f1f7d0476fd
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.519 seconds |


Install hpccsystems-platform-community_6.4.4-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented:
@richardkchapman - good to merge.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: cf7cee154ffeb231efc754b41fc97e65626a1782
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.242 seconds |


Install hpccsystems-platform-community_6.4.4-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
@jakesmith have you checked that this change does not introduce any possibility of deadlock?

jakesmith commented:
@richardkchapman - yes, it looks fine.
